utils/char16trie/src/char16trie.rs

```diff
@@ -0,0 +1,396 @@
+///    The value, if present, is for the string/char16_t sequence so far.
+///
+///  Aside from the value, which uses the node lead unit's high bits:
+///
+///  - Linear-match node: Matches a number of units.
+///  - Branch node: Branches to other nodes according to the current input unit.
+///    The node unit is the length of the branch (number of units to select from)
+///    minus 1. It is followed by a sub-node:
+///    - If the length is at most MAX_BRANCH_LINEAR_SUB_NODE_LENGTH, then
+///      there are length-1 (key, value) pairs and then one more comparison unit.
+///      If one of the key units matches, then the value is either a final value for
+///      the string so far, or a "jump" delta to the next node.
+///      If the last unit matches, then matching continues with the next node.
+///      (Values have the same encoding as final-value nodes.)
+///    - If the length is greater than MAX_BRANCH_LINEAR_SUB_NODE_LENGTH, then
+///      there is one unit and one "jump" delta.
+///      If the input unit is less than the sub-node unit, then "jump" by delta to
+///      the next sub-node which will have a length of length/2.
+///      (The delta has its own compact encoding.)
+///      Otherwise, skip the "jump" delta to the next sub-node
+///      which will have a length of length-length/2.
+///
+/// For more information:
+/// - [ICU4C UCharsTrie](https://unicode-org.github.io/icu-docs/apidoc/released/icu4c/classicu_1_1UCharsTrie.html)
+/// - [ICU4J CharsTrie](https://unicode-org.github.io/icu-docs/apidoc/released/icu4j/com/ibm/icu/util/CharsTrie.html) API.
+#[cfg_attr(feature = "serde", derive(Serialize, Deserialize))]
+#[derive(Clone)]
+pub struct Char16Trie<'data> {
+    /// An array of u16 containing the trie data.
+    #[cfg_attr(feature = "serde", serde(borrow))]
+    pub data: ZeroVec<'data, u16>,
+}
+
+/// This struct represents an iterator over a Char16Trie.
+pub struct Char16TrieIterator<'a> {
+    /// A reference to the Char16Trie data to iterate over.
+    trie: &'a [<u16 as zerovec::ule::AsULE>::ULE],
+    /// Index of next trie unit to read, or None if there are no more matches.
+    pos: Option<usize>,
+    /// Remaining length of a linear-match node, minus 1, or None if not in
+    /// such a node.
+    remaining_match_length: Option<usize>,
+}
+
+#[derive(Clone, Copy, Debug, PartialEq)]
+pub enum TrieResult {
+    /// The input unit(s) did not continue a matching string.
+    /// Once next() return TrieResult::NoMatch, all further calls to next()
+    /// will also return TrieResult::NoMatch.
+    NoMatch,
+    /// The input unit(s) matched a string but there is no value for the string
+    /// so far.  (It is a prefix of a longer string.)
+    NoValue,
+    /// The input unit(s) continued a matching string and there is a value for
+    /// the string so far. No further input byte/unit can continue a matching
+    /// string.
+    FinalValue(i32),
```

Comment:
Question: where does the type `i32` come from?  Why is it signed?  Does the data structure support other values?  Does it support storing a u16 or u8 to save on space?  @markusicu

Reply:
I used `i32` because that is what the ICU4C implementation uses for `getValue`: https://github.com/unicode-org/icu/blob/main/icu4c/source/common/unicode/ucharstrie.h#L258

Reply:
The data structure is defined as mapping sequences of uint16_t to int32_t, leaving it up to the user to interpret the values. Some don't care about the value at all, only about the TrieResult. In many cases, these will be small-ish integers. For collation, it will be interpreted as a uint32_t. (AFAIR the internal encoding favors non-negative values, essentially storing uint32_t in a variable-length way -- but programmers like to use signed integers by default.)